src/main.rs

```diff
@@ -226,5 +276,99 @@ fn main() {
+            manifest_type,
+            manifest_version,
+            input,
+            output,
+        } => {
+            let (mut input, mut output) = open_files(input, output);
+
+            let key = fs::read(key).expect("failed to open file");
+            let keypair = ring::rsa::Keypair::from_pkcs8(&key)
+                .expect("failed to parse key");
+            let mut signer = ring::rsa::Builder::new()
+                .new_signer(keypair)
+                .expect("failed to create signing engine");
+
+            let mut manifest_buf = vec![0; 0x2000];
+            let mut container = Containerizer::new(&mut manifest_buf)
+                .expect("failed to create containerizer")
+                .with_type(manifest_type)
+                .expect("failed to set manifest type")
+                .with_metadata(&Metadata {
+                    version_id: manifest_version,
+                })
+                .expect("failed to set manifest metadata");
+
+            match manifest_type {
+                ManifestType::Fpm => {
+                    let mut buf = Vec::new();
+                    input.read_to_end(&mut buf).expect("failed to read file");
+
+                    let fpm: Fpm<provenance::Adhoc> =
+                        serde_json::from_slice(&buf)
+                            .expect("failed to parse JSON");
+                    fpm.unparse(&mut container)
+                        .expect("failed to serialize manifest");
+                }
+            }
+
+            let manifest = container
+                .sign(&mut signer)
+                .expect("failed to sign manifest");
+            output
+                .write_all(manifest)
+                .expect("failed to write manifest");
+        }
+        CliCommand::ShowManifest {
+            key,
+            pretty,
+            input,
+            output,
+        } => {
+            let (mut input, output) = open_files(input, output);
+
+            let engine = key.map(|key| {
+                let key = fs::read(key).expect("failed to open file");
+                let keypair = ring::rsa::Keypair::from_pkcs8(&key)
+                    .expect("failed to parse key");
+                ring::rsa::Builder::new()
+                    .new_engine(keypair.public())
+                    .expect("failed to create signature verification engine")
+            });
+            let mut buf = Vec::new();
+            input.read_to_end(&mut buf).expect("failed to read file");
+
+            let container = match engine {
+                Some(mut engine) => {
+                    match Container::parse_and_verify(&buf, &mut engine) {
+                        Ok(c) => c,
+                        Err(manifest::Error::SignatureFailure) => {
+                            panic!("failed to verify signature")
+                        }
+                        Err(_) => panic!("failed to parse manifest container"),
+                    }
+                    .downgrade()
```

Comment:
Why does this need to be downgraded?

Reply:
To make type unification across branches work out. We're going to immediately serialize it so the provenance doesn't matter.